### ISSUE TITLE
Shutdown issue

### ISSUE THREAD

**Issue by myrnazhou** [cefsharp/CefSharp#585]
Hi guys,
I can't seems to call the Shutdown, I think this case is similar with this one #2. 
So the problem is I can't close my app because when I called Shutdown the thread stuck there, and the app just continue running.
https://github.com/cefsharp/CefSharp/blob/12f9c297e859948f307b7daceba9577c448c69f6/CefSharp.Core/Cef.h#L357

**Comment by Bodekaer** [comment]
Please re-open this issue, as it still occurs. I have the exact same popup as @myrnazhou.

**Comment by Bodekaer** [comment]
I see the same popup as @myrnazhou when closing the application, and running CEF Shutdown.
It's happening randomly, and hard to reproduce.

@myrnazhou did you try @jornh's suggestion to first dispose all browsers before calling shutdown?

**Comment by amaitland** [comment]
@Bodekaer Has the issue occurred since you updated to the latest version of `Cef 2171`?

Possibly useful links
- http://stackoverflow.com/questions/99552/where-do-pure-virtual-function-call-crashes-come-from/100555#100555
- http://www.artima.com/cppsource/pure_virtual.html
- http://www.lenholgate.com/blog/2006/01/purecall.html

**Comment by Bodekaer** [comment]
Update: 
I tried the suggestion from #2 to dispose all browsers and call `GC.Collect()` and it appears to have solved it!
Thanks a lot for the tips.

I believe the main cause was that we sometimes store browsers in memory without displaying them (for faster re-load times when displaying new browsers).
I just added a simple script to dispose all those browsers before calling `Shutdown()`.

Wonderful! :) This was one that's been worrying me for a while, and yet it was so easy to fix :) hehe

**Comment by amaitland** [comment]
@Bodekaer Excellent! Is it worth implementing the changes in `CefSharp`?

**Comment by Bodekaer** [comment]
Yeah I guess it would. Although I'm not so happy with my implementation yet, so wouldn't feel too good about putting it into CefSharp :)
Anyways, here is the simple quick solution:


     public class ChromiumWebBrowser

         private static readonly List<WebControl> _instances = new List<WebControl>();

         public ChromiumWebBrowser(){
               _instances.Add(this);
         }

        protected override void Dispose(bool isdisposing)
        {
            _instances.Remove(this);
            base.Dispose(isdisposing);
        }

        public static void DisposeAllInstances()
        {
            // Loop through all instances and remove.
            while (_instances.Count > 0)
            {
                _instances[0].Dispose();
            }
        }

    }

And then I just call `ChromiumWebBrowser.DisposeAllInstances()` before calling `Cef.Shutdown()`.
I'm sure this can be done 100x better, just haven't had time to look more into it yet.
:)

**Comment by amaitland** [comment]
Interesting! I'm surprised that's nessicary. See
https://github.com/cefsharp/CefSharp/blob/cef/2171/CefSharp.Core/Cef.h#L330

Already does something very similar.

**Comment by Bodekaer** [comment]
Haha yes that is strange :)
Well it seems like it doesn't call the `Dispose()` here, which may be the cause, as we have certain event handlers that are unsubscribed in the `Disposing(bool isDisposing)` method of our extended `ChromiumWebBrowser`.

Could that be the reason?

**Comment by amaitland** [comment]
> Well it seems like it doesn't call the Dispose() here, which may be the cause, as we have certain event handlers that are unsubscribed in the Disposing(bool isDisposing) method of our extended ChromiumWebBrowser

Quite possible, probably worth looking into :+1:

**Comment by amaitland** [comment]
@Bodekaer Happy to close this?

**Comment by Bodekaer** [comment]
I actually just experienced this again yesterday. 
It seems to still happen in very rare cases, but I think it's more related to cases where the JS execution somehow timed out, which seems to causes the dispose calls to freeze.

Let's close this issue though, as once I have a way to reproduce this other rare case, then I'll just open a new issue. 
Since I've only experienced this one scenario once, it might also just be caused by something else and not be CefSharp related.
